content/en/docs/concepts/extend-kubernetes/compute-storage-net/resource-drivers.md

```diff
@@ -0,0 +1,288 @@
+
+
+### Kubelet-plugin registration
+
+Resource driver kubelet-plugin has to implement [NodeServer interface](https://pkg.go.dev/k8s.io/kubelet@v0.28.2/pkg/apis/dra/v1alpha3#NodeServer),
+which can be simply used with [kubelet-plugin helper instance](https://github.com/kubernetes-sigs/dra-example-driver/blob/151b7c8da2e620c47a3591e1a937f8d0297b0c25/cmd/dra-example-kubeletplugin/main.go#L160).
+You can use any programming language, but the examples are all written in Go.
+
+## Controller implementation
+
+A resource driver controller's main responsibility is to allocate and deallocate resources for
+{{< glossary_tooltip term_id="resource-claim" text="ResourceClaims">}}.
+
+There are two modes of allocation the ResourceClaim can have:
+
+- `WaitForFirstConsumer` (default), AKA "delayed". In this mode the resource will only be requested
+  for ResourceClaim when a Pod that needs it is being scheduled.
+- `Immediate`: the resource has to be allocated to the ResourceClaim as soon as possible, and
+  retained until the ResourceClaim is deleted.
+
+If `allocationMode` is not set explicitly, the mode is `WaitForFirstConsumer`.
+
+### Delayed allocation
+
+Delayed allocation is done in two steps: first, the scheduler picks up a list of potential Node names
+and asks from resource driver controller which of them are unsuitable for ResourceClaim through
+_UnsuitableNodes_ call. If no nodes were suitable, scheduler selects another batch of Node names for
+the same call, until suitable node is found.
+
+After at least one Node is found to be suitable for all ResourceClaims, the scheduler
+considers the suitable nodes for the other Pod scheduling constraints (native resources
+requests, affinity, selectors, etc.), and picks up exactly one Node name. The resource driver
+controller is then asked to allocate resources for the ResourceClaim on the selected node through
+_Allocate_ call. All ResourceClaims, that Pod references and that match the resource driver name in
+ResourceClass, will be passed as parameters into this call.
+
+If Allocate call returns error for any number of ResourceClaims, the helper code will repeat the
+same call with interval until it succeeds.
+
+### Immediate allocation
+
+Immediate allocation does not have selected node, and it is up to the Resource Driver controller
+to select the best suitable node based on the ResourceClaim, ResourceClass and their parameters.
+Therefore in this scenario the `UnsupportedNodes` is not called, but only `Allocate`.
+
+## Common calls for both allocation modes
+
+In both modes the allocation is preceded by getting parameters objects for ResourceClaim and
+ResourceClass to ensure the Resource Driver is able to get these objects and understand them.
+
+
+```gRPC
+    // GetClassParameters gets called to retrieve the parameter object
```

Comment:
IMHO all these would read better with "is called" instead of "Get... gets called".

Reply:
This is quote from the code, we don't fix it in here.

Reply:
Our code wasn't reviewed for language the same way as this docs PR now is. We can and should consider the advice we get here and make the same change in code PR.

Reply:
All right.

Reply:
I'm wondering whether we should link to the [gopkg page](https://pkg.go.dev/k8s.io/dynamic-resource-allocation@v0.28.3/controller) instead of duplicating the API documentation here.

That API might still change, even from one release to the next. We don't want to continuously update this page just to keep it in sync.

Reply:
If this is an API that we expect end users and integrators to use, we should document it.

If we don't have capacity to do that, or expect not to, that's fine but that lack of maintenance capacity should be something we consider as part of approving graduation to GA.

Reply:
This is not an API for end users, nor for integrators (= people setting up a cluster?).

We don't mirror the API for client-go here either, do we? This API falls into the same "for developers" category.

Reply:
Our official client for end users to interact with the Kubernetes HTTP API is `kubectl`, but we do have a page about [client libraries](https://kubernetes.io/docs/reference/using-api/client-libraries/). Of course, you don't need to use `client-go` at all and many people choose not to.

What do we have / will we have for people who are not part of the Kubernetes project and want to write a resource driver?

Reply:
Moved to reference.